weed/s3api/s3api_circuit_breaker.go

```diff
@@ -9,23 +10,78 @@ import (
 	"github.com/chrislusf/seaweedfs/weed/pb/s3_pb"
 	"github.com/chrislusf/seaweedfs/weed/s3api/s3_constants"
 	"github.com/chrislusf/seaweedfs/weed/s3api/s3err"
+	"github.com/chrislusf/seaweedfs/weed/stats"
 	"github.com/gorilla/mux"
 	"net/http"
-	"sync"
+	"strconv"
+	"strings"
 	"sync/atomic"
 )
 
+type TrafficLimits map[string]*BucketTrafficLimit
```

Comment:
how about using a struct as such instead of a map here?
```
type TrafficLimits {
  requestCount  int64
  throughputMB int64
}
```

Reply:
> how about using a struct as such instead of a map here?
> 
> ```
> type TrafficLimits {
>   requestCount  int64
>   throughputMB int64
> }
> ```

没这样做是因为目前 bucket 限流的 action 和 limitType 的值有多个，limitType 有两种：`Count` 和 `MB`，但是`action` 可能是 `Read`,`Write`,`Admin`,`Tagging`，这样下来 action 和 limitType 的组合会比较多；所以如果采用这种方式的话，需要在代码里面就会针对不同的 action 判断取不同的成员变量，如果action判断多了，会让代码显得乱；

不过实际限流用到的就只有 `Read` 和 `Write`，要不就只支持这两种类型的限流；对于读（下载）请求，只有请求数 `Count` 的限制，对于写请求，支持并发 请求数和 吞吐量的的限制，这样的话，结构体变成下面这样：
```go
type TrafficLimits struct {
	readRequestCountTraffic  *BucketTrafficLimit
	writeRequestCountTraffic *BucketTrafficLimit
	writeThroughputTraffic   *BucketTrafficLimit
}

type BucketTrafficLimit struct {
	max   int64
	count int64
}
```

你怎么看 @chrislusf